app/models/concerns/orchestration/external_ipam.rb

```diff
@@ -0,0 +1,145 @@
+module Orchestration::ExternalIPAM
+  extend ActiveSupport::Concern
+  include Orchestration::Common
+  include SubnetsHelper
+
+  included do
+    after_validation :queue_external_ipam
+    before_destroy :queue_external_ipam_destroy
+  end
+
+  def generate_external_ipam_task_id(action, network_type, interface = self)
+    id = [interface.mac, interface.ip, interface.identifier, interface.id].find {|x| x&.present?}
+    "external_ipam_#{action}_#{id}_#{network_type}"
+  end
+
+  protected
+
+  def set_add_external_ip(params)
+    ip, subnet = params[:ip], params[:subnet]
+
+    if ip_is_available?(ip, subnet)
+      subnet.external_ipam_proxy.add_ip_to_subnet(ip, subnet.network_address)
+    else
+      self.errors.add :ip, _('This IP address has already been reserved in External IPAM') if subnet.network_type == "IPv4"
+      self.errors.add :ip6, _('This IP address has already been reserved in External IPAM') if subnet.network_type == "IPv6"
+      self.errors.add :interfaces, _('Some interfaces are invalid')
+    end
+  end
+
+  def set_remove_external_ip(params)
+    ip, subnet = params[:ip], params[:subnet]
+    subnet.external_ipam_proxy.delete_ip_from_subnet(ip, subnet.network_address)
+  end
+
+  def del_add_external_ip(params)
+    new_record? ? rollback_on_add : rollback_on_update
+  end
+
+  def del_remove_external_ip(params)
+  end
+
+  def rollback_on_add
```

Comment:
This is to handle the edge case where IPv4 and IPv6 addresses are suggested, then user manually changes one of the addresses(lets say IPv4) to an address that is already taken. When user clicks submit, error is thrown saying "IPv4 is taken", but the IPv6 job succeeds resulting in persisting it to IPAM, even though Host creation has errored out. 

Then, if the user corrects the IPv4 to an available address, and submits again, the IPv6 address will show as taken. Not a good UI/UX. This logic rolls back only when there is an error on one of the addresses. If IPv4 errors out, IPv6 is rolled back and vice versa. 

I couldn't really see any other way of doing this from within the orchestration, but happy to implement another way, if one exists

Reply:
A rollback should put all systems involved in the transaction back to clean state. I'd expect that the IP address which was successfully created must be deleted too. Even if this means it can't be reused again or any other reason.

Reply:
I still don't see the case handled where there's both an IPv4 and IPv6 error

Reply:
Both IPv4 and IPv6 are handled in separate transactions, so whether there is one error, or an error on both, it is handled. When one fails, the other is rolled back. i.e. when IPv4 fails, IPv6 is rolled back, and vice-versa